### PR TITLE
Allow to generate several samples for each prompt

### LINKED ISSUE #416: Restore the ability to draw multiple samples with Open Source models
This was removed in #366 to simplify the PR, and should be added again. This will require to be careful with the shape and an added dimension will need to be added to take sample shape into account. The mechanism implemented there can be re-used when implementing Beam Search #258

### PR DESCRIPTION
**Pull request by rlouf**
Closes #416 

### TODO

- [x] Use `torch.repeat_interleave` so samples of the same sequence in a batch are contiguous (easier for beam search)
- [x] Make sure the FSMs are corrrectly duplicated

### Questions

- Do we make the sequence generator return arrays of shape `(num_samples * batch_size, num_tokens)`, and reshape in `SequenceGenerator`? Reshaping in the generator makes the latter much more complex. We could also not reshape at all, and let the user do it manually, which will simplify chained calls in the future.

### CODE REVIEW

File: outlines/generate/generator.py [hunk 1]
```diff
@@ -78,6 +79,17 @@ def sequence_generator(
...
     token_ids, attention_masks, kv_cache = init_state
+    batch_shape = token_ids.shape[:-1]
+
+    # To take several samples we duplicate `token_ids`, `attention_masks`
+    # and `fsm_states` as many times as the number of samples requested.
+    # The resulting tensors are of shape (num_samples * num_batches, num_tokens)
+    token_ids = torch.repeat_interleave(token_ids, num_samples, dim=0)
+    attention_masks = torch.repeat_interleave(attention_masks, num_samples, dim=0)
+
+    fsm_states = fsm_states * num_samples
```
Comment (rlouf) [inline outlines/generate/generator.py:90]: Repeat in the same way we repeat `fsms`

File: outlines/generate/api.py [hunk 2]
```diff
@@ -225,8 +242,9 @@ def __call__(
                 break
 
         generated_token_ids = self.get_generated_token_ids(
```
Comment (rlouf) [inline outlines/generate/api.py:244]: Include `tokenizer.decode` in this function to make the code more understandable.

File: outlines/generate/api.py [hunk 3]
```diff
@@ -225,8 +242,9 @@ def __call__(
...
-            init_state, prompts, last_state
+            init_state, prompts, last_state, num_samples
         )
+
         generated = self.tokenizer.decode(generated_token_ids)
         stripped = [
```
Comment (rlouf) [inline outlines/generate/api.py:249]: Iterate over sequences inside the function for readability

File: outlines/generate/api.py [hunk 4]
```diff
@@ -225,8 +242,9 @@ def __call__(
...
             self.strip_stop_sequences(sequence, stop_sequences)
```
Comment (rlouf) [inline outlines/generate/api.py:250]: Same, also for `format_sequence` (and catch the exception inside)

**Review by lapp0 — COMMENTED** [review]
I'm still learning the `outlines/generate/` side of the code-base. If there are any components you would like test cases for, that would help me learn :)

IMHO we shouldn't reshape at all. Users should expect an array of results with the same length as their input prompts. A shape-agnostic decoder should be a final step so we aren't required to convert to decodable array and back again in multiple places.

File: outlines/generate/api.py [hunk 5]
```diff
@@ -14,16 +14,26 @@
...
+        *,
+        max_tokens=None,
+        stop_at=None,
+    ):
         self.generate_token = token_generator(model, sampler)
         self.fsm = fsm
         self.tokenizer = model.tokenizer
         self.device = device
         self.max_tokens = max_tokens
+        self.num_particles = sampler.particles
```
[the remark this answers is not in the document]
Comment (rlouf) [reply outlines/generate/api.py:36]: I'm trying to find a word that works for multinomial, beam search and SMC :)

File: outlines/generate/api.py [hunk 6]
```diff
@@ -60,18 +71,28 @@ def get_generated_token_ids(
...
-        prompt_lengths = [len(prompt_token_ids[i]) for i in range(len(prompts))]
+        prompt_lengths = [
+            len(prompt_token_ids[i])
+            for _ in range(num_samples)
+            for i in range(len(prompts))
+        ]
+
+        # We flatten the obtained token_ids since the tokenizer's decoder
+        # only accepts tensor with two dimensions
+        token_ids = last_state.token_ids.reshape((-1, last_state.token_ids.shape[-1]))
```
[the remark this answers is not in the document]
Comment (rlouf) [reply outlines/generate/api.py:91]: How would that work? Currently the `sequence_generator` yields arrays of shape `(n_samples, n_batches, n_tokens)` but `batch_decode` will only accept 2-dimensional arrays.

File: outlines/generate/generator.py [hunk 7]
```diff
@@ -55,7 +58,8 @@ def sequence_generator(
     fsms: List["FSM"],
     init_state: Tuple,
     fsm_states: List[FSMState],
-    rng: torch.Generator,
+    num_samples: int = 1,
```
Comment (rlouf) [inline outlines/generate/generator.py:61]: While I think `sequence_generator` should be able to take inputs of any shape, I am not sure it should be aware of `num_samples`.

File: outlines/generate/api.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lapp0) [reply outlines/generate/api.py:91]: I'm suggesting  a wrapper for `batch_decode` which reshapes any N-dim input into a 2-dim format for decoding, then reverts the decoded output back to its original dimensions to be returned.

This means we can maintain the structure of our data throughout the process

[the remark this answers is not in the document]
Comment (rlouf) [reply outlines/generate/api.py:91]: The issue with this is that we need to modify every post-processing function. I am currently leaning towards handling the multiple samples outside of `generators`, which would simplify the code a lot.
